Civi/Standalone/WebEntrypoint.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+namespace Civi\Standalone;
+
+/**
+ * Logic for loading web pages (would be in index.php, but here we can update it)
+ */
+class WebEntrypoint {
+
+  /**
+   * If we think Civi is installed, run the page
+   *
+   * Otherwise show the Web installer
+   *
+   * @see self::checkCiviInstalled
+   */
+  public static function index(): void {
+    if (self::checkCiviInstalled()) {
+      ErrorHandler::setHandler();
+      self::invoke();
+    }
+    else {
+      self::installer();
+    }
+  }
+
+  /**
+   * if CIVICRM_INSTALLED flag is set in settings, we think the database
+   * *should* already be installed and we'll never show the installer
+   *
+   * if its not set, we check for presence of an existing database
+   *
+   * in general setting the flag is probably a good idea to never show the installer again
+   * e.g. you dont want to show the installer if your database goes down
+   * (especially as the standalone installer is permissionless - no cms user accounts! - and
+   * may even know your database credentials if these are provided as env variables)
+   */
+  public static function checkCiviInstalled(): bool {
+    if (defined('CIVICRM_INSTALLED')) {
+      return !!CIVICRM_INSTALLED;
+    }
+    if (defined('CIVICRM_DSN')) {
+      // can we autodetect a good value here?
+    }
```

Comment:
So the reason this is here: [in the follow up PR](https://github.com/civicrm/civicrm-core/pull/30132), the idea is you *can* provide a settings file (and/or env vars) prior to running the installer, with DSN settings in them.

At this point the signal is not whether there is a settings file, but whether the CIVICRM_INSTALLED is set to TRUE in the settings. (The installer writes an additional settings file into the settings directory which adds this line).

To my mind it would be nice if CIVICRM_INSTALLED isn't defined to be able to fallback to checking any provided DSN details to see if they point to a valid install, and not show the installer in this case. But I was struggling to find an existing way to do this at this point in the bootstrap.

Reply:
I can leave out until it's actually functioning though, sure.

Reply:
(commented out)

Reply:
1. The PR has comments about security / flaky connections / re-installation -- that's a really good point.

1. I relate to the idea of checking DSN -- but ultimately agree that we should *not* check it here/now.
    * There's a scope mismatch. On one side, you have a simple boolean (`WebEntrypoint::checkCiviInstalled()`). On the other side, "installation" refers to a process with various resources, phases, failure-points, future-features. Those are not the same size/complexity. Intuitively, there will be cases where the boolean is an anemic representation of the true installation-status.
    * Aside: There exists `$setup->checkInstalled()`. This actually returns an object which can hold more details. But you're right to *not* use that here. It does more work, and (for 99.9% of pageloads) you don't need that work.
    * There are hypotheticals for more complex things, but I don't think they're important here/now, and I vote to keep the check simple+predictable. (And we're not locked in -- the whole point of this file is to make it easier to change/evolve.)

1. On a gut level, `CIVICRM_INSTALLED` doesn't feel like a configuration value. It kinda feels like repeating the mistakes of  `set_include_path()` and `CRM_Core_ClassLoader::singleton()->register();`. (They're oddballs in the settings file which aren't proper configuration data -- but once you've propagated the template to some large subset of sites, it's PITA to cleanup.*)
    * In other UF's, if you wanted to identify whether you had active/loaded settings, you could check `CIVICRM_SETTINGS_PATH`. (*This value is set right before loading the file; eg [d7](https://github.com/civicrm/civicrm-drupal/blob/7.x-master/civicrm.module#L186-L188), [d8+](https://github.com/civicrm/civicrm-drupal-8/blob/c3988a297bbe852f23a682c66472480dcb549320/src/Civicrm.php#L40-L45), [joomla](https://github.com/civicrm/civicrm-joomla/blob/6287dd5d3614b2f5cedf9081126aa363ad96d4e9/site/civicrm.php#L10-L11).*)
    * I think this would keep the settings-file tidier (*lower maintenance*) and align with other UFs: https://github.com/civicrm/civicrm-core/commit/c309e697f49ede3452dd0bd02fb36cd04c27c210

Reply:
> In other UF's, if you wanted to identify whether you had active/loaded settings, you could check CIVICRM_SETTINGS_PATH. (This value is set right before loading the file; eg [d7](https://github.com/civicrm/civicrm-drupal/blob/7.x-master/civicrm.module#L186-L188), [d8+](https://github.com/civicrm/civicrm-drupal-8/blob/c3988a297bbe852f23a682c66472480dcb549320/src/Civicrm.php#L40-L45), [joomla](https://github.com/civicrm/civicrm-joomla/blob/6287dd5d3614b2f5cedf9081126aa363ad96d4e9/site/civicrm.php#L10-L11).)

So I was deliberately trying to avoid checking CIVICRM_SETTINGS_PATH. If a boolean is an anemic representation of installation status, then the settings path is a bad proxy for the boolean.

At the moment we *need* a boolean, because there's a binary choice: do we show the installer or not when you hit index.php.

I didn't want to use SETTINGS_PATH, because A) I want to be able to provide a settings file before installation, so an installer (or SAS hoster or ...) can provide some configuration for the installer to use. B) I think if we had good enough defaults, you could might legitimately run a site _without_ a settings file.

I was basically thinking of CIVICRM_INSTALLED as a "show the web installer?" settings - and it's kind of a tertiary:
TRUE => never show the installer; FALSE => show the installer; NULL = "AUTO" => determine whether to show the installer based on whether we can find an installed database. For the concerns above, AUTO isn't an ideal value.. but I can see it having uses.

From that perspective, I don't see the parallels with `set_include_path` / registering the classloader. It feels like legitimate config state to me.

Reply:
> At the moment we need a boolean, because there's a binary choice: do we show the installer or not when you hit index.php.

This did make me think... what if we have a dedicated url path for the installer `/install` ? Then we can jump straight to `invoke` if we're not on that path. I still think we need a way to check whether *not* to show the installer on a site that is already installed, which I would see taking the same route

Reply:
( For other UFs, wouldn't the ideal check would be "is_module_enabled('civicrm')" ? Though I can see that being complicated if the CiviCRM installation doesn't fit entirely within the "enable a module" paradigm of the UF. )

Reply:
> I don't understand the idea of running without a settings file. The db credentials gotta go somewhere, right?

The hope is to provide with ENV vars. 

And I also think there's lots of legitimate use cases for having a settings file before install, in order to provide some settings to the installer. 

For both those reasons I think an explicit variable makes more sense than using CIVICRM_SETTINGS_PATH directly. 

> On the other side, "installation" refers to a process with various resources, phases, failure-points, future-features. Those are not the same size/complexity. Intuitively, there will be cases where the boolean is an anemic representation of the true installation-status.

Isn't this even more of a problem with `defined('CIVICRM_SETTINGS_PATH')`?

> Would renaming » CIVICRM_INSTALLED » CIVICRM_SHOW_INSTALLER be clearer?

 :shrug: I can see advantages either way but I think " the installer outputs a settings file which sets `CIVICRM_INSTALLED = 1`" makes more sense than "the installer has to unset CIVICRM_SHOW_INSTALLER".